sdk/endpoint/ethernet.go

```diff
@@ -0,0 +1,75 @@
+package endpoint
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes/empty"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/connectioncontext"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/networkservice"
+)
+
+//NewEthernetEndpoint creates a new endpoint that adds ethernet context with specific mac addresses
+func NewEthernetEndpoint(srcMac, dstMac string) networkservice.NetworkServiceServer {
+	return &ethernet{srcMac, dstMac}
+}
+
+type ethernet struct {
+	dstMac string
+	srcMac string
+}
+
+func (m *ethernet) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
+	return &empty.Empty{}, nil
+}
+
+func (m *ethernet) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+	c := request.GetConnection()
+	if c.GetContext() != nil {
+		c.GetContext().EthernetContext = &connectioncontext.EthernetContext{}
+		if m.srcMac != "" {
+			c.GetContext().GetEthernetContext().SrcMacAddress = m.srcMac
+		}
+		if m.dstMac != "" {
+			c.GetContext().GetEthernetContext().DstMacAddress = m.dstMac
+		} else {
+			ip := ""
+			if c.GetContext().GetIpContext() != nil {
+				ip = strings.Split(c.Context.IpContext.DstIpAddr, "/")[0]
+			}
+			c.GetContext().GetEthernetContext().DstMacAddress = generateMacBaseOnIP(ip)
+		}
+		if err := c.GetContext().GetEthernetContext().Validate(); err != nil {
+			return request.GetConnection(), err
+		}
+	}
+	return request.GetConnection(), nil
+}
+
+func (m *ethernet) Name() string {
+	return "Ethernet Mutator Endpoint"
+}
+
+func generateMacBaseOnIP(ip string) string {
+	ipAddr := net.ParseIP(ip).To4()
+	mac := administeredAddress([]byte(ipAddr))
+	return net.HardwareAddr(mac).String()
+}
+
+//http://www.noah.org/wiki/MAC_address#locally_administered_address
+func administeredAddress(input []byte) []byte {
+	result := make([]byte, 6)
+	or := []byte{2, 0, 0, 0, 0, 0}
+	and := []byte{254, 255, 255, 255, 255, 255}
+	min := len(result)
+	if min > len(input) {
+		min = len(input)
+	}
+	for i := 0; i < min; i++ {
+		result[i] = or[i] | input[i]&and[i]
+	}
+	return result
+}
```

Comment:
This is very good if I want to specify the Ethernet Address.  But often, I don't.  We just want to capture the Ethernet Address generated for the interface by VPP and set it as the dstMac in the Connection returned from the Request.

Reply:
@edwarnicke This is clear for me but in the proposal  https://github.com/networkservicemesh/networkservicemesh/issues/1563 we should apply mac addresses from an endpoint only for memif connections. (Please, correct me if this is wrong)

What we should do if dataplane received a request with ethernet context with not empty mac addresses and the connection has kernel interface mechanism?

My options are:
1) Ignore these addresses and rewrite fields uses vpp agent generated mac address
2) Throw error
3) Your option